src/geode/model/helpers/ray_tracing.cpp

```diff
@@ -32,6 +32,55 @@
 #include <geode/model/mixin/core/surface.h>
 #include <geode/model/representation/core/brep.h>
 
+namespace
+{
+    geode::BoundarySurfaceIntersections find_intersections_with_boundaries(
+        const geode::Ray3D& ray,
+        const geode::BRep& brep,
+        const geode::Block3D& block )
+    {
+        geode::BoundarySurfaceIntersections result;
+        for( const auto& surface : brep.boundaries( block ) )
+        {
+            const auto aabb = geode::create_aabb_tree( surface.mesh() );
+            geode::RayTracing3D ray_tracing{ surface.mesh(), ray };
+            aabb.compute_ray_element_bbox_intersections( ray, ray_tracing );
+            result[surface.id()] = ray_tracing.all_intersections();
+        }
+        return result;
+    }
+
+    absl::optional< geode::index_t > count_real_intersections_with_boudaries(
+        const geode::Ray3D& ray,
+        const geode::BRep& brep,
+        const geode::Block3D& block )
+    {
+        geode::index_t nb_intersections{ 0 };
+        const auto tracing =
+            find_intersections_with_boundaries( ray, brep, block );
+        for( const auto& intersections : tracing )
+        {
+            for( const auto& intersection : intersections.second )
+            {
+                if( intersection.position != geode::Position::inside )
```

Comment:
If intersection point is on surface triangle edge or vertex, what is the expected behavior?

Reply:
Any advise? the problem is that if the intersection is on a vertex or an edge it is ambiguous in this context.

Reply:
maybe a solution is to slightly change the direction of the ray untill we get a non ambigous response. the risk of infinit loop looks small don't you think?. 
we can either  chose the direction randomly or rotate the direction by a chosen increment.